File: packages/typescript/src/features/trust/services/TrustEngine.ts
import { logger } from "../../../logger.ts";
import {
	type Component,
	type IAgentRuntime,
	type Metadata,
	Service,
	type UUID,
} from "../../../types/index.ts";
import { stringToUuid } from "../../../utils.ts";

import {
	type TrustCalculationConfig,
	type TrustContext,
	type TrustDecision,
	type TrustDimensions,
	type TrustEvidence,
	TrustEvidenceType,
	type TrustInteraction,
	type TrustProfile,
	type TrustRequirements,
} from "../types/trust.ts";

import { getDb } from "./db.ts";
import { getTrustEvidence, insertTrustEvidence } from "./SecurityStore.ts";

/**
 * Default configuration for trust calculations
 */
const DEFAULT_CONFIG: TrustCalculationConfig = {
	recencyBias: 0.7,
	evidenceDecayRate: 0.5, // Points per day
	minimumEvidenceCount: 3,
	verificationMultiplier: 1.5,
	dimensionWeights: {
		reliability: 0.25,
		competence: 0.2,
		integrity: 0.25,
		benevolence: 0.2,
		transparency: 0.1,
	},
};

/**
 * Evidence impact mapping for different evidence types
 */
const EVIDENCE_IMPACT_MAP: Record<
	TrustEvidenceType,
	{ dimensions: Partial<TrustDimensions>; baseImpact: number }
> = {
	// Positive evidence
	[TrustEvidenceType.PROMISE_KEPT]: {
		dimensions: { reliability: 15, integrity: 10 },
		baseImpact: 10,
	},
	[TrustEvidenceType.HELPFUL_ACTION]: {
		dimensions: { benevolence: 15, competence: 10 },
		baseImpact: 8,
	},
	[TrustEvidenceType.CONSISTENT_BEHAVIOR]: {
		dimensions: { reliability: 20, transparency: 10 },
		baseImpact: 12,
	},
	[TrustEvidenceType.VERIFIED_IDENTITY]: {
		dimensions: { transparency: 20, integrity: 10 },
		baseImpact: 15,
	},
	[TrustEvidenceType.COMMUNITY_CONTRIBUTION]: {
		dimensions: { benevolence: 20, competence: 15 },
		baseImpact: 12,
	},
	[TrustEvidenceType.SUCCESSFUL_TRANSACTION]: {
		dimensions: { reliability: 15, competence: 15 },
		baseImpact: 10,
	},

	// Negative evidence
	[TrustEvidenceType.PROMISE_BROKEN]: {
		dimensions: { reliability: -25, integrity: -15 },
		baseImpact: -15,
	},
	[TrustEvidenceType.HARMFUL_ACTION]: {
		dimensions: { benevolence: -30, integrity: -20 },
		baseImpact: -20,
	},
	[TrustEvidenceType.INCONSISTENT_BEHAVIOR]: {
		dimensions: { reliability: -20, transparency: -15 },
		baseImpact: -12,
	},
	[TrustEvidenceType.SUSPICIOUS_ACTIVITY]: {
		dimensions: { integrity: -15, transparency: -20 },
		baseImpact: -15,
	},
	[TrustEvidenceType.FAILED_VERIFICATION]: {
		dimensions: { transparency: -25, integrity: -10 },
		baseImpact: -10,
	},
	[TrustEvidenceType.SPAM_BEHAVIOR]: {
		dimensions: { benevolence: -15, competence: -10 },
		baseImpact: -10,
	},
	[TrustEvidenceType.SECURITY_VIOLATION]: {
		dimensions: { integrity: -35, reliability: -20 },
		baseImpact: -25,
	},

	// Neutral evidence
	[TrustEvidenceType.IDENTITY_CHANGE]: {
		dimensions: { transparency: -5 },
		baseImpact: 0,
	},
	[TrustEvidenceType.ROLE_CHANGE]: {
		dimensions: {},
		baseImpact: 0,
	},
	[TrustEvidenceType.CONTEXT_SWITCH]: {
		dimensions: {},
		baseImpact: 0,
	},
};

export class TrustEngine extends Service {
	static serviceType = "trust-engine" as const;

	capabilityDescription =
		"Multi-dimensional trust scoring and evaluation system";

	private static readonly ACTION_CONTEXT_WEIGHTS: Record<
		string,
		Partial<TrustCalculationConfig["dimensionWeights"]>
	> = {
		financial: {
			integrity: 0.35,
			reliability: 0.3,
			competence: 0.15,
			benevolence: 0.1,
			transparency: 0.1,
		},
		moderation: {
			benevolence: 0.3,
			integrity: 0.25,
			competence: 0.2,
			reliability: 0.15,
			transparency: 0.1,
		},
		content_creation: {
			competence: 0.3,
			integrity: 0.2,
			reliability: 0.2,
			benevolence: 0.15,
			transparency: 0.15,
		},
	};

	private trustConfig: TrustCalculationConfig;
	private profileCache: Map<string, TrustProfile> = new Map();
	private readonly cacheTimeout = 5 * 60 * 1000; // 5 minutes
	private readonly maxInteractionsInMemory = 500;
	private interactions: TrustInteraction[] = [];
	private rateLimits: Map<
		string,
		{ count: number; windowStart: number; typeHistory: Map<string, number> }
	> = new Map();
	private readonly maxEvidencePerHour = 10;

	constructor(config?: Partial<TrustCalculationConfig>) {
		super();
		this.trustConfig = { ...DEFAULT_CONFIG, ...config };
	}

	async initialize(runtime: IAgentRuntime): Promise<void> {
		this.runtime = runtime;
		logger.info(
			{ trustConfig: this.trustConfig },
			"[TrustEngine] Initialized with config:",
		);
	}

	async stop(): Promise<void> {
		this.profileCache.clear();
		logger.info("[TrustEngine] Stopped");
	}

	static async start(runtime: IAgentRuntime): Promise<Service> {
		const service = new TrustEngine();
		await service.initialize(runtime);
		return service;
	}

	/**
	 * Calculate trust profile for an entity
	 */
	async calculateTrust(
		subjectId: UUID,
		context: TrustContext,
	): Promise<TrustProfile> {
		const cacheKey = `${context.evaluatorId}-${subjectId}`;

		// Check cache
		const cached = this.profileCache.get(cacheKey);
		if (cached && Date.now() - cached.lastCalculated < this.cacheTimeout) {
			return cached;
		}

		// Load evidence from components
		const evidence = await this.loadEvidence(subjectId, context);

		// Calculate dimensions
		const dimensions = this.calculateDimensions(evidence);

		// Resolve dimension weights (context-specific overrides or defaults)
		const baseWeights = this.trustConfig.dimensionWeights;
		const overrides = context.action
			? TrustEngine.ACTION_CONTEXT_WEIGHTS[context.action]
			: undefined;
		const activeWeights = overrides
			? {
					reliability: overrides.reliability ?? baseWeights.reliability,
					competence: overrides.competence ?? baseWeights.competence,
					integrity: overrides.integrity ?? baseWeights.integrity,
					benevolence: overrides.benevolence ?? baseWeights.benevolence,
					transparency: overrides.transparency ?? baseWeights.transparency,
				}
			: baseWeights;

		// Calculate overall trust with resolved weights
		const overallTrust = this.calculateOverallTrust(dimensions, activeWeights);

		// Calculate confidence
		const confidence = this.calculateConfidence(evidence);

		// Analyze trend
		const trend = await this.analyzeTrend(subjectId, context, overallTrust);

		const profile: TrustProfile = {
			entityId: subjectId,
			dimensions,
			overallTrust,
			confidence,
			interactionCount: evidence.length,
			evidence: evidence.slice(0, 100), // Keep most recent 100
			lastCalculated: Date.now(),
			calculationMethod: "dimensional_aggregation_v1",
			trend,
			evaluatorId: context.evaluatorId,
		};

		// Save to cache and storage
		this.profileCache.set(cacheKey, profile);
		await this.saveTrustProfile(profile, context);

		return profile;
	}

	/**
	 * Records a trust interaction
	 */
	async recordInteraction(interaction: TrustInteraction): Promise<void> {
		// Rate limit check
		const rateCheck = this.checkRateLimit(
			interaction.targetEntityId,
			interaction.type,
		);
		if (!rateCheck.allowed) {
			logger.warn(
				{ entityId: interaction.targetEntityId, type: interaction.type },
				"[TrustEngine] Rate limit exceeded, skipping interaction recording",
			);
			return;
		}

		// Apply diminishing returns weight to impact
		if (rateCheck.weight < 1.0) {
			interaction.impact = interaction.impact * rateCheck.weight;
		}

		this.interactions.push(interaction);
		// Trim in-memory interactions to prevent unbounded growth
		if (this.interactions.length > this.maxInteractionsInMemory) {
			this.interactions = this.interactions.slice(
				-this.maxInteractionsInMemory,
			);
		}

		// Invalidate cache entries for the affected entity
		for (const key of this.profileCache.keys()) {
			if (key.endsWith(`-${interaction.targetEntityId}`)) {
				this.profileCache.delete(key);
			}
		}

		logger.info(
			{
				type: interaction.type,
				impact: interaction.impact,
				source: interaction.sourceEntityId,
				target: interaction.targetEntityId,
			},
			"[TrustEngine] Recorded interaction:",
		);

		// Persist to database
		try {
			const db = getDb(this.runtime);
			await insertTrustEvidence(db, {
				targetEntityId: interaction.targetEntityId,
				sourceEntityId: interaction.sourceEntityId,
				evaluatorId: interaction.context?.evaluatorId || this.runtime.agentId,
				type: interaction.type,
				impact: interaction.impact,
				weight: rateCheck.weight,
				description: interaction.details?.description || "",
				verified: true,
				context: (interaction.context ?? {}) as Record<string, unknown>,
			});
		} catch (err: unknown) {
			logger.warn(
				{ error: err },
				"[TrustEngine] Failed to persist trust evidence to DB",
			);
		}
	}

	/**
	 * Evaluate if an action is allowed based on trust
	 */
	async evaluateTrustDecision(
		entityId: UUID,
		requirements: TrustRequirements,
		context: TrustContext,
	): Promise<TrustDecision> {
		const profile = await this.calculateTrust(entityId, context);

		// Check overall trust
		if (profile.overallTrust < requirements.minimumTrust) {
			return {
				allowed: false,
				trustScore: profile.overallTrust,
				requiredScore: requirements.minimumTrust,
				dimensionsChecked: profile.dimensions,
				reason: `Trust score ${profile.overallTrust} is below required ${requirements.minimumTrust}`,
				suggestions: this.generateTrustBuildingSuggestions(
					profile,
					requirements,
				),
			};
		}

		// Check specific dimensions
		if (requirements.dimensions) {
			for (const [dimension, required] of Object.entries(
				requirements.dimensions,
			)) {
				const actual = profile.dimensions[dimension as keyof TrustDimensions];
				if (actual < required) {
					return {
						allowed: false,
						trustScore: profile.overallTrust,
						requiredScore: requirements.minimumTrust,
						dimensionsChecked: requirements.dimensions,
						reason: `${dimension} score ${actual} is below required ${required}`,
						suggestions: this.generateDimensionSuggestions(
							dimension as keyof TrustDimensions,
						),
					};
				}
			}
		}

		// Check interaction count
		if (
			requirements.minimumInteractions &&
			profile.interactionCount < requirements.minimumInteractions
		) {
			return {
				allowed: false,
				trustScore: profile.overallTrust,
				requiredScore: requirements.minimumTrust,
				dimensionsChecked: profile.dimensions,
				reason: `Insufficient interactions: ${profile.interactionCount} < ${requirements.minimumInteractions}`,
				suggestions: ["Engage in more interactions to build history"],
			};
		}

		// Check confidence
		if (
			requirements.minimumConfidence &&
			profile.confidence < requirements.minimumConfidence
		) {
			return {
				allowed: false,
				trustScore: profile.overallTrust,
				requiredScore: requirements.minimumTrust,
				dimensionsChecked: profile.dimensions,
				reason: `Trust confidence ${profile.confidence} is below required ${requirements.minimumConfidence}`,
				suggestions: [
					"More consistent interactions needed to increase confidence",
				],
			};
		}

		return {
			allowed: true,
			trustScore: profile.overallTrust,
			requiredScore: requirements.minimumTrust,
			dimensionsChecked: profile.dimensions,
			reason: "All trust requirements met",
		};
	}

	/**
	 * Check rate limiting for evidence recording
	 */
	private static readonly DIMINISHING_WEIGHTS = [1.0, 0.75, 0.5, 0.25] as const;

	private checkRateLimit(
		entityId: UUID,
		evidenceType: TrustEvidenceType,
	): { allowed: boolean; weight: number } {
		const now = Date.now();
		const hourMs = 60 * 60 * 1000;

		let entry = this.rateLimits.get(entityId);
		if (!entry) {
			entry = { count: 0, windowStart: now, typeHistory: new Map() };
			this.rateLimits.set(entityId, entry);
		}

		// Reset window if more than 1 hour has passed
		if (now - entry.windowStart > hourMs) {
			entry.count = 0;
			entry.windowStart = now;
			entry.typeHistory.clear();
			// Prune stale entries from other entities while we're here
			for (const [key, val] of this.rateLimits) {
				if (now - val.windowStart > hourMs * 2) this.rateLimits.delete(key);
			}
		}

		if (entry.count >= this.maxEvidencePerHour) {
			return { allowed: false, weight: 0 };
		}

		// Diminishing returns: 1st occurrence = 1.0, 2nd = 0.75, 3rd = 0.5, 4th+ = 0.25
		const typeCount = entry.typeHistory.get(evidenceType) ?? 0;
		const weight =
			TrustEngine.DIMINISHING_WEIGHTS[
				Math.min(typeCount, TrustEngine.DIMINISHING_WEIGHTS.length - 1)
			];

		entry.count++;
		entry.typeHistory.set(evidenceType, typeCount + 1);

		return { allowed: true, weight };
	}

	/**
	 * Calculate trust dimensions from evidence
	 */
	private calculateDimensions(evidence: TrustEvidence[]): TrustDimensions {
		const dimensions: TrustDimensions = {
			reliability: 50,
			competence: 50,
			integrity: 50,
			benevolence: 50,
			transparency: 50,
		};

		for (const ev of evidence) {
			const impact = EVIDENCE_IMPACT_MAP[ev.type];
			if (!impact) continue;

			// Apply age weight
			const ageWeight = this.calculateAgeWeight(ev.timestamp);

			// Apply verification multiplier
			const verificationMultiplier = ev.verified
				? this.trustConfig.verificationMultiplier
				: 1.0;

			// Update dimensions
			for (const [dimension, value] of Object.entries(impact.dimensions)) {
				const adjustedValue =
					value * ev.weight * ageWeight * verificationMultiplier;
				dimensions[dimension as keyof TrustDimensions] = Math.max(
					0,
					Math.min(
						100,
						dimensions[dimension as keyof TrustDimensions] + adjustedValue,
					),
				);
			}
		}

		return dimensions;
	}

	/**
	 * Calculate overall trust score from dimensions
	 */
	private calculateOverallTrust(
		dimensions: TrustDimensions,
		weights: TrustCalculationConfig["dimensionWeights"] = this.trustConfig
			.dimensionWeights,
	): number {
		let weightedSum = 0;
		let totalWeight = 0;

		for (const [dimension, value] of Object.entries(dimensions)) {
			const weight = weights[dimension as keyof TrustDimensions];
			weightedSum += value * weight;
			totalWeight += weight;
		}

		return Math.round(weightedSum / totalWeight);
	}

	/**
	 * Calculate confidence based on evidence quantity and consistency
	 */
	private calculateConfidence(evidence: TrustEvidence[]): number {
		if (evidence.length < this.trustConfig.minimumEvidenceCount) {
			return 0;
		}

		// Base confidence from evidence count
		const countConfidence = Math.min(1, evidence.length / 20);

		// Consistency factor - how consistent is the evidence?
		const positiveCount = evidence.filter((e) => e.impact > 0).length;
		const negativeCount = evidence.filter((e) => e.impact < 0).length;
		const consistency =
			1 - Math.abs(positiveCount - negativeCount) / evidence.length;

		// Recency factor - how recent is the evidence?
		const recentEvidence = evidence.filter(
			(e) => Date.now() - e.timestamp < 7 * 24 * 60 * 60 * 1000, // 7 days
		);
		const recencyFactor = recentEvidence.length / evidence.length;

		return countConfidence * 0.4 + consistency * 0.3 + recencyFactor * 0.3;
	}

	/**
	 * Calculate age weight for evidence based on recency
	 */
	private calculateAgeWeight(timestamp: number): number {
		const ageInDays = (Date.now() - timestamp) / (24 * 60 * 60 * 1000);
		const decayFactor = Math.exp(
			-this.trustConfig.evidenceDecayRate * ageInDays,
		);

		// Blend with recency bias
		return (
			this.trustConfig.recencyBias * decayFactor +
			(1 - this.trustConfig.recencyBias) * 0.5
		);
	}

	/**
	 * Analyze trust trend over time
	 */
	private async analyzeTrend(
		entityId: UUID,
		context: TrustContext,
		currentScore: number,
	): Promise<TrustProfile["trend"]> {
		// Load historical trust scores
		const components = await this.runtime.getComponents(entityId);
		const historicalProfiles = components
			.filter(
				(c) => c.type === "trust_profile" && c.agentId === context.evaluatorId,
			)
			.map((c) => c.data as unknown as TrustProfile)
			.sort((a, b) => b.lastCalculated - a.lastCalculated)
			.slice(0, 10);

		if (historicalProfiles.length < 2) {
			return {
				direction: "stable",
				changeRate: 0,
				lastChangeAt: Date.now(),
			};
		}

		// Calculate trend
		const previousScore = historicalProfiles[0].overallTrust;
		const oldestScore =
			historicalProfiles[historicalProfiles.length - 1].overallTrust;
		const timeSpanDays =
			(Date.now() -
				historicalProfiles[historicalProfiles.length - 1].lastCalculated) /
			(24 * 60 * 60 * 1000);

		const changeRate = (currentScore - oldestScore) / timeSpanDays;

		let direction: "increasing" | "decreasing" | "stable";
		if (Math.abs(changeRate) < 0.5) {
			direction = "stable";
		} else if (changeRate > 0) {
			direction = "increasing";
		} else {
			direction = "decreasing";
		}

		return {
			direction,
			changeRate: Math.round(changeRate * 10) / 10,
			lastChangeAt:
				currentScore !== previousScore
					? Date.now()
					: historicalProfiles[0].trend?.lastChangeAt || Date.now(),
		};
	}

	/**
	 * Load evidence from storage
	 */
	private async loadEvidence(
		entityId: UUID,
		context: TrustContext,
	): Promise<TrustEvidence[]> {
		const components = await this.runtime.getComponents(entityId);

		const evidenceComponents = components.filter(
			(c) =>
				c.type === "trust_evidence" &&
				(!context.worldId || c.worldId === context.worldId) &&
				(!context.roomId || c.roomId === context.roomId),
		);

		const evidence: TrustEvidence[] = [];
		for (const component of evidenceComponents) {
			const ev = component.data as unknown as TrustEvidence;

			// Apply time window filter
			if (context.timeWindow) {
				if (
					ev.timestamp < context.timeWindow.start ||
					ev.timestamp > context.timeWindow.end
				) {
					continue;
				}
			}

			evidence.push(ev);
		}

		// Also load from DB and merge (deduplicate by timestamp+type)
		try {
			const db = getDb(this.runtime);
			const dbRows = await getTrustEvidence(db, entityId, context.evaluatorId);
			const existingKeys = new Set(
				evidence.map((e) => `${e.timestamp}-${e.type}`),
			);
			for (const row of dbRows) {
				const rowTimestamp =
					row.timestamp instanceof Date
						? row.timestamp.getTime()
						: (row.timestamp as number);
				const key = `${rowTimestamp}-${row.type as string}`;
				if (!existingKeys.has(key)) {
					evidence.push({
						type: row.type as TrustEvidenceType,
						timestamp: rowTimestamp,
						impact: row.impact as number,
						weight: (row.weight as number) ?? 1.0,
						description: (row.description as string) ?? "",
						reportedBy: row.sourceEntityId as UUID,
						targetEntityId: row.targetEntityId as UUID,
						verified: (row.verified as boolean) ?? false,
						context: (row.context as TrustContext) ?? {
							evaluatorId: context.evaluatorId,
						},
						evaluatorId: (row.evaluatorId as UUID) ?? context.evaluatorId,
					});
					existingKeys.add(key);
				}
			}
		} catch (err: unknown) {
			logger.warn(
				{ error: err },
				"[TrustEngine] Failed to load trust evidence from DB",
			);
		}

		return evidence.sort((a, b) => b.timestamp - a.timestamp);
	}

	/**
	 * Save trust profile to storage
	 */
	private async saveTrustProfile(
		profile: TrustProfile,
		context: TrustContext,
	): Promise<void> {
		const componentId = stringToUuid(
			`trust-profile-${profile.entityId}-${context.evaluatorId}`,
		);

		const worldId = context.worldId || stringToUuid("trust-world");
		await this.runtime.ensureWorldExists({
			id: worldId,
			name: "trust-world",
			agentId: this.runtime.agentId,
			messageServerId: stringToUuid("default"),
			metadata: {},
		});

		// Check if component already exists
		const existingComponent = await this.runtime.getComponent(
			profile.entityId,
			"trust_profile",
			worldId,
			context.evaluatorId,
		);

		const component: Component = {
			id: componentId,
			type: "trust_profile",
			agentId: context.evaluatorId,
			entityId: profile.entityId,
			roomId: context.roomId || stringToUuid("trust-global"),
			worldId,
			sourceEntityId: context.evaluatorId,
			data: profile as unknown as Metadata,
			createdAt: existingComponent?.createdAt || Date.now(),
		};

		if (existingComponent) {
			// Update existing component
			await this.runtime.updateComponent(component);
		} else {
			// Create new component
			await this.runtime.createComponent(component);
		}
	}

	/**
	 * Generate suggestions for building trust
	 */
	private generateTrustBuildingSuggestions(
		profile: TrustProfile,
		requirements: TrustRequirements,
	): string[] {
		const suggestions: string[] = [];

		// Overall trust suggestions
		if (profile.overallTrust < requirements.minimumTrust) {
			const gap = requirements.minimumTrust - profile.overallTrust;
			suggestions.push(
				`Build ${gap} more trust points through positive interactions`,
			);
		}

		// Dimension-specific suggestions
		const weakestDimension = Object.entries(profile.dimensions).sort(
			([, a], [, b]) => a - b,
		)[0][0];

		suggestions.push(
			...this.generateDimensionSuggestions(
				weakestDimension as keyof TrustDimensions,
			),
		);

		// Interaction count suggestions
		if (profile.interactionCount < 10) {
			suggestions.push("Engage in more conversations and activities");
		}

		return suggestions;
	}

	/**
	 * Generate suggestions for improving specific dimensions
	 */
	private generateDimensionSuggestions(
		dimension: keyof TrustDimensions,
	): string[] {
		const suggestions: Record<keyof TrustDimensions, string[]> = {
			reliability: [
				"Keep your promises and commitments",
				"Be consistent in your actions",
				"Follow through on what you say",
			],
			competence: [
				"Demonstrate your skills through helpful contributions",
				"Share valuable knowledge or resources",
				"Complete tasks successfully",
			],
			integrity: [
				"Be honest and transparent in your communications",
				"Admit mistakes when they happen",
				"Follow community guidelines consistently",
			],
			benevolence: [
				"Help other community members",
				"Show genuine interest in others' wellbeing",
				"Contribute positively to discussions",
			],
			transparency: [
				"Be open about your intentions",
				"Share information freely when appropriate",
				"Verify your identity on multiple platforms",
			],
		};

		return (
			suggestions[dimension] || [
				"Continue building trust through positive interactions",
			]
		);
	}

	/**
	 * Evaluates trust for an entity (simplified API for actions)
	 */
	async evaluateTrust(
		entityId: UUID,
		evaluatorId: UUID,
		context?: Partial<TrustContext>,
	): Promise<TrustProfile> {
		const fullContext: TrustContext = {
			evaluatorId,
			...context,
		};
		return this.calculateTrust(entityId, fullContext);
	}

	/**
	 * Get recent trust interactions for an entity
	 * @param daysBack Number of days to look back (default: 10)
	 */
	async getRecentInteractions(
		entityId: UUID,
		daysBack = 10,
	): Promise<TrustInteraction[]> {
		const cutoff = Date.now() - daysBack * 24 * 60 * 60 * 1000;
		return this.interactions.filter(
			(i) =>
				(i.sourceEntityId === entityId || i.targetEntityId === entityId) &&
				i.timestamp > cutoff,
		);
	}
}

File: packages/rust/src/features/core_capabilities/trust/trust_engine.rs
//! TrustEngineService — calculates and manages trust profiles.

use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use uuid::Uuid;

use super::types::*;

/// Service that calculates and manages trust profiles for entities.
pub struct TrustEngineService {
    profiles: Arc<RwLock<HashMap<Uuid, TrustProfile>>>,
    config: TrustCalculationConfig,
}

impl TrustEngineService {
    /// Create a new TrustEngineService with default configuration.
    pub fn new() -> Self {
        Self {
            profiles: Arc::new(RwLock::new(HashMap::new())),
            config: TrustCalculationConfig::default(),
        }
    }

    /// Create with custom configuration.
    pub fn with_config(config: TrustCalculationConfig) -> Self {
        Self {
            profiles: Arc::new(RwLock::new(HashMap::new())),
            config,
        }
    }

    /// Get a trust profile for an entity.
    pub async fn get_profile(&self, entity_id: Uuid) -> Option<TrustProfile> {
        self.profiles.read().await.get(&entity_id).cloned()
    }

    /// Record a trust interaction and update the profile.
    pub async fn record_interaction(
        &self,
        interaction: TrustInteraction,
    ) -> anyhow::Result<TrustProfile> {
        let now = chrono::Utc::now().timestamp_millis();
        let evaluator_id = interaction
            .context
            .as_ref()
            .map(|c| c.evaluator_id)
            .unwrap_or_default();

        let evidence = TrustEvidence {
            evidence_type: interaction.interaction_type.clone(),
            timestamp: interaction.timestamp,
            impact: interaction.impact,
            weight: 1.0,
            description: interaction
                .details
                .as_ref()
                .and_then(|d| d.get("description"))
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            reported_by: interaction.source_entity_id,
            verified: false,
            context: interaction.context.clone().unwrap_or_default(),
            target_entity_id: interaction.target_entity_id,
            evaluator_id,
            metadata: interaction.details.clone(),
        };

        let mut profiles = self.profiles.write().await;
        let profile = profiles
            .entry(interaction.target_entity_id)
            .or_insert_with(|| TrustProfile {
                entity_id: interaction.target_entity_id,
                dimensions: TrustDimensions::default(),
                overall_trust: 50.0,
                confidence: 0.0,
                interaction_count: 0,
                evidence: Vec::new(),
                last_calculated: now,
                calculation_method: "weighted_average".to_string(),
                trend: TrustTrend {
                    direction: TrustTrendDirection::Stable,
                    change_rate: 0.0,
                    last_change_at: now,
                },
                evaluator_id,
            });

        profile.evidence.push(evidence);
        profile.interaction_count += 1;

        // Recalculate trust
        self.recalculate_profile(profile);

        Ok(profile.clone())
    }

    /// Recalculate trust dimensions and overall score.
    fn recalculate_profile(&self, profile: &mut TrustProfile) {
        let now = chrono::Utc::now().timestamp_millis();
        let old_trust = profile.overall_trust;

        // Calculate weighted evidence impact per dimension
        let mut dim_scores = [0.0f64; 5]; // reliability, competence, integrity, benevolence, transparency
        let mut dim_weights = [0.0f64; 5];

        for evidence in &profile.evidence {
            let age_days = (now - evidence.timestamp) as f64 / (24.0 * 60.0 * 60.0 * 1000.0);
            let decay = (-age_days * self.config.evidence_decay_rate / 100.0).exp();
            let recency_weight =
                self.config.recency_bias * decay + (1.0 - self.config.recency_bias);
            let verification_weight = if evidence.verified {
                self.config.verification_multiplier
            } else {
                1.0
            };

            let total_weight = evidence.weight * recency_weight * verification_weight;
            let impact = evidence.impact * total_weight;

            // Map evidence types to dimensions
            match &evidence.evidence_type {
                TrustEvidenceType::PromiseKept | TrustEvidenceType::ConsistentBehavior => {
                    dim_scores[0] += impact;
                    dim_weights[0] += total_weight;
                }
                TrustEvidenceType::HelpfulAction | TrustEvidenceType::SuccessfulTransaction => {
                    dim_scores[1] += impact;
                    dim_weights[1] += total_weight;
                }
                TrustEvidenceType::VerifiedIdentity | TrustEvidenceType::CommunityContribution => {
                    dim_scores[2] += impact;
                    dim_weights[2] += total_weight;
                }
                TrustEvidenceType::PromiseBroken | TrustEvidenceType::InconsistentBehavior => {
                    dim_scores[0] += impact; // negative impact
                    dim_weights[0] += total_weight;
                }
                TrustEvidenceType::HarmfulAction | TrustEvidenceType::SecurityViolation => {
                    dim_scores[2] += impact;
                    dim_weights[2] += total_weight;
                    dim_scores[3] += impact;
                    dim_weights[3] += total_weight;
                }
                TrustEvidenceType::SpamBehavior | TrustEvidenceType::SuspiciousActivity => {
                    dim_scores[4] += impact;
                    dim_weights[4] += total_weight;
                }
                _ => {
                    // Neutral evidence - minor impact on transparency
                    dim_scores[4] += impact * 0.1;
                    dim_weights[4] += total_weight * 0.1;
                }
            }
        }

        // Normalize dimensions to 0-100 scale
        let base_score = 50.0;
        for i in 0..5 {
            if dim_weights[i] > 0.0 {
                dim_scores[i] = (base_score + dim_scores[i] / dim_weights[i]).clamp(0.0, 100.0);
            } else {
                dim_scores[i] = base_score;
            }
        }

        profile.dimensions = TrustDimensions {
            reliability: dim_scores[0],
            competence: dim_scores[1],
            integrity: dim_scores[2],
            benevolence: dim_scores[3],
            transparency: dim_scores[4],
        };

        // Calculate overall trust
        let w = &self.config.dimension_weights;
        profile.overall_trust = (profile.dimensions.reliability * w.reliability
            + profile.dimensions.competence * w.competence
            + profile.dimensions.integrity * w.integrity
            + profile.dimensions.benevolence * w.benevolence
            + profile.dimensions.transparency * w.transparency)
            .clamp(0.0, 100.0);

        // Calculate confidence
        let evidence_count = profile.evidence.len();
        profile.confidence = if evidence_count >= self.config.minimum_evidence_count {
            (evidence_count as f64 / (evidence_count as f64 + 10.0)).min(1.0)
        } else {
            evidence_count as f64 / self.config.minimum_evidence_count as f64 * 0.5
        };

        // Update trend
        let change = profile.overall_trust - old_trust;
        profile.trend = TrustTrend {
            direction: if change > 1.0 {
                TrustTrendDirection::Increasing
            } else if change < -1.0 {
                TrustTrendDirection::Decreasing
            } else {
                TrustTrendDirection::Stable
            },
            change_rate: change,
            last_change_at: now,
        };

        profile.last_calculated = now;
    }

    /// Make a trust-based decision.
    pub async fn check_trust(
        &self,
        entity_id: Uuid,
        requirements: &TrustRequirements,
    ) -> TrustDecision {
        let profile = self.profiles.read().await.get(&entity_id).cloned();

        match profile {
            Some(profile) => {
                let allowed = profile.overall_trust >= requirements.minimum_trust
                    && requirements
                        .minimum_confidence
                        .map(|min| profile.confidence >= min)
                        .unwrap_or(true)
                    && requirements
                        .minimum_interactions
                        .map(|min| profile.interaction_count >= min)
                        .unwrap_or(true);

                let reason = if allowed {
                    format!(
                        "Trust score {:.1} meets requirement of {:.1}",
                        profile.overall_trust, requirements.minimum_trust
                    )
                } else {
                    format!(
                        "Trust score {:.1} below requirement of {:.1}",
                        profile.overall_trust, requirements.minimum_trust
                    )
                };

                let suggestions = if !allowed {
                    Some(vec![
                        "Build trust through consistent, helpful interactions".to_string(),
                        "Verify your identity to boost trust score".to_string(),
                    ])
                } else {
                    None
                };

                TrustDecision {
                    allowed,
                    trust_score: profile.overall_trust,
                    required_score: requirements.minimum_trust,
                    dimensions_checked: profile.dimensions,
                    reason,
                    suggestions,
                }
            }
            None => TrustDecision {
                allowed: requirements.minimum_trust <= 0.0,
                trust_score: 0.0,
                required_score: requirements.minimum_trust,
                dimensions_checked: TrustDimensions::default(),
                reason: "No trust profile exists for this entity".to_string(),
                suggestions: Some(vec![
                    "Start interacting to build a trust profile".to_string()
                ]),
            },
        }
    }
}

impl Default for TrustEngineService {
    fn default() -> Self {
        Self::new()
    }
}
